pkg/api/api_test.go

```diff
@@ -336,6 +338,73 @@ func TestPostageHeaderError(t *testing.T) {
 	}
 }
 
+// TestPostageDirectAndDeferred tests that incorrect postage batch ids
+// provided to the api correct the appropriate error code.
+func TestPostageDirectAndDeferred(t *testing.T) {
+	var (
+		mockStorer               = mock.NewStorer()
+		mockStatestore           = statestore.NewStateStore()
+		logger                   = logging.New(io.Discard, 5)
+		mp                       = mockpost.New(mockpost.WithIssuer(postage.NewStampIssuer("", "", batchOk, big.NewInt(3), 11, 10, 1000, true)))
+		client, _, _, chanStorer = newTestServer(t, testServerOptions{
+			Storer:       mockStorer,
+			Tags:         tags.NewTags(mockStatestore, logger),
+			Logger:       logger,
+			Post:         mp,
+			DirectUpload: true,
+		})
+
+		endpoints = []string{
+			"bytes", "bzz", "chunks",
+		}
+	)
+	for _, endpoint := range endpoints {
+		t.Run(endpoint+": deferred", func(t *testing.T) {
+			hexbatch := hex.EncodeToString(batchOk)
+			chunk := testingc.GenerateTestRandomChunk()
+			var responseBytes []byte
+			jsonhttptest.Request(t, client, http.MethodPost, "/"+endpoint, http.StatusCreated,
+				jsonhttptest.WithRequestHeader(api.SwarmPostageBatchIdHeader, hexbatch),
+				jsonhttptest.WithRequestHeader(api.ContentTypeHeader, "application/octet-stream"),
+				jsonhttptest.WithRequestHeader(api.SwarmDeferredUploadHeader, "true"),
+				jsonhttptest.WithRequestBody(bytes.NewReader(chunk.Data())),
+				jsonhttptest.WithPutResponseBody(&responseBytes),
+			)
+			var body struct {
+				Reference swarm.Address `json:"reference"`
+			}
+			if err := json.Unmarshal(responseBytes, &body); err != nil {
+				t.Fatal("unmarshal response body:", err)
+			}
+			if found, _ := mockStorer.Has(context.Background(), body.Reference); !found {
+				t.Fatal("chunk not found in the store")
+			}
```

Comment:
i would also check that the chan store does not have the chunk (for whatever reason). the inverse also in the next test